core/src/HIP/Kokkos_HIP_Instance.hpp

```diff
@@ -65,6 +65,93 @@ HIP::size_type *hip_internal_scratch_flags(const HIP &instance,
 
 //----------------------------------------------------------------------------
 
+// Helper to protect a shared resource from being used by multiple streams
+// simultaneously.
+// If used properly, only one stream at a time will be able to use the shared
+// resource.
+// This helper should be used in a thread-safe way.
+//
+// Typical usage:
+// @code
+// auto lock_for_acquisition = shared_resource.acquire();
+//
+// ... do stuff on the shared resource ...
+//
+// shared_resource.release(std::move(lock_for_acquisition), stream);
+//
+// ...
+//
+// auto lock_for_involvement = shared_resource.lock();
+// hipStreamSynchronize(stream);
+// shared_resource.check_if_involved(std::move(lock_for_involvement), stream);
+// hipStreamDestroy(stream);
+// @endcode
+struct SharedResourceLocking {
```

Comment:
The functionality looks more like an event manager.
```suggestion
struct HIPEventManager {
```

Reply:
I'm not sure that's a better name. When I see 'event' and 'manager', I’d expect something like a centralized event factory or similar. Also, the 'locking' aspect seems pretty important—shouldn't that be reflected in the name?

Reply:
Fair enough. I can also see the perspective that the event is an implementation detail (that influences the behavior so that the underlying event is destroyed before the stream).
Maybe, `SharedResourceLock` is still a better name since the interface is somewhat close to C++ lock classes.

Reply:
OK, I've changed the name to `SharedResourceLock` :heavy_check_mark: